api/controller.py

```diff
@@ -647,7 +667,13 @@ def load_work(self, library, identifier_type, identifier):
 
         # We know there is at least one LicensePool, and all LicensePools
         # for an Identifier have the same Work.
-        return pools[0].work
+        work = pools[0].work
+
+        if work and not work.age_appropriate_for_patron(self.request_patron):
+            # This work is not age-appropriate for the authenticated
+            # patron. Don't show it.
+            work = NOT_AGE_APPROPRIATE
```

Comment:
Would it make sense to change the current approach and instead of returning `ProblemDetail` objects from functions and handling them with `if isinstance(result, ProblemDetail)` statements just raise a `ProblemDetail` exception handling it at the bottom of the callstack where we could log them appropriately and return a `ProblemDetail` object back to the caller if needed? Of course, I don't imply that it's needed to be done in this PR but maybe we can consider it for the future? I think it could simplify the codebase, we could get rid of a lot of `if isinstance(result, ProblemDetail)` statements

Reply:
I have had similar thoughts in the past and I think it could work. The main thing we'd have to change is situations where one piece of code calls `ProblemDetail.detailed` to add more information to a `ProblemDetail` object. I think on balance the code would be easier to understand if we made this change.